robotframework-ls/src/robotframework_ls/impl/flow_explorer_model_builder.py

```diff
@@ -1,9 +1,53 @@
+from contextlib import contextmanager
 from robotframework_ls.impl.protocols import ICompletionContext, IKeywordFound, INode
 from typing import List, Any, Dict, Tuple
 import os
 from robocorp_ls_core.basic import isinstance_name
 
 
+class _KeywordRecursionStack:
+    _stack: Dict
+
+    def __init__(self) -> None:
+        self._stack = {}
+
+    def has(self, keyword_name: str) -> bool:
+        return keyword_name in self._stack
+
+    def push(self, keyword_name: str) -> None:
+        self._stack.update({keyword_name: None})
```

Comment:
The idiom for this with a dict would be:

`self._stack[keyword_name] = None`

Still if you change the stack for a `set` it'd be `self._stack.add(keyword_name)`.

Additional notes:
- since `pop` would remove the entry, I think you should assert that you don't double-push.
- since this `push` should only be used inside the class it should be protected (or it could be inlined).

i.e.:

```
def _push(self, keyword_name: str):
    assert keyword_name not in self._stack
    self._stack.add(keyword_name)
```